utils/s3/client.cc

```diff
@@ -854,6 +861,209 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+        }
+
+        co_await _bg_uploads.close();
+        std::exception_ptr ex;
+        try {
+            co_await complete_multipart_upload();
+        } catch (...) {
+            ex = std::current_exception();
+        }
+        if (ex) {
+            co_await abort_multipart_upload();
+            std::rethrow_exception(ex);
+        }
+    }
+
+    future<> put_object(file&& f, uint64_t len) {
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html
+        s3l.trace("PUT {} ({})", _object_name, _path.native());
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        if (_tag) {
+            req._headers["x-amz-tagging"] = seastar::format("{}={}", _tag->key, _tag->value);
+        }
+        req.write_body("bin", len, [f = std::move(f), this] (output_stream<char>&& out_) mutable -> future<> {
+            auto output = std::move(out_);
+            auto input = make_file_input_stream(std::move(f));
+            co_await input.consume(throttled_consumer{output, *_client});
+            co_await output.close();
+            co_await input.close();
+        });
+        co_await _client->make_request(std::move(req), [len, start = s3_clock::now()] (group_client& gc, const auto& rep, auto&& in) {
+            gc.write_stats.update(len, s3_clock::now() - start);
+            return ignore_reply(rep, std::move(in));
+        });
+    }
+
+public:
+    do_upload_file(shared_ptr<client> cln,
+                   std::filesystem::path path,
+                   sstring object_name,
+                   std::optional<tag> tag,
+                   size_t part_size)
+        : _client{std::move(cln)}
+        , _path{std::move(path)}
+        , _object_name{std::move(object_name)}
+        , _tag{std::move(tag)}
+        , _part_size{part_size} {
+    }
+
+    future<> upload() {
+        auto f = co_await open_file_dma(_path.native(), open_flags::ro);
+        const auto stat = co_await f.stat();
+        const uint64_t file_size = stat.st_size;
+        // use multipart upload when possible in order to transmit parts in
+        // parallel to improve throughput
+        if (file_size > aws_minimum_part_size) {
+            size_t part_size = calc_part_size(file_size, _part_size);
+            co_await multi_part_upload(std::move(f), file_size, part_size);
+        } else {
+            // single part upload
+            co_await put_object(std::move(f), file_size);
+        }
+    }
+};
+
+future<> client::upload_file(std::filesystem::path path,
+                              sstring object_name,
+                              std::optional<tag> tag,
+                              std::optional<size_t> part_size) {
+    do_upload_file do_upload{shared_from_this(),
+                             std::move(path),
+                             std::move(object_name),
+                             std::move(tag),
+                             part_size.value_or(0)};
+    co_await do_upload.upload();
```

Comment:
Just out of curiosity -- why introducing the class do_upload_file instead of writing most of the code right in here?

Reply:
i wanted 

1. to model the multipart upload using some self-contained structure, so it's obvious to posterity that "this" and "that" are for multipart upload for file. and
2. to make sure the data dependency is obvious

ideally, i should put something like (in pseudo code):
```c++
auto upload_id = co_await create_multipart_upload(object_name);
/// could use generator!
for (auto part : upload_parts(f, part_size)) {
  co_await upload_part(upload_id, std::move(part));
}
try {
  co_await complete_multipart_upload(upload_id);
} catch (exception& e) {
  co_await abort_multipart_upload();
  throw;
}
```
so, yes, it can be implemented using free functions. the reason it's implemented this way was to group these helper together, but i can surely try to implement these as functions after addressing the other comments.